meilisearch-lib/src/index/search.rs

```diff
@@ -436,50 +445,71 @@ fn make_document(
+    // recreate the original json
+    for (key, value) in obkv.iter() {
+        let value = serde_json::from_slice(value)?;
+        let key = field_ids_map
+            .name(key)
+            .expect("Missing field name")
+            .to_string();
 
-            // This unwrap must be safe since we got the ids from the fields_ids_map just
-            // before.
-            let key = field_ids_map
-                .name(*attr)
-                .expect("Missing field name")
-                .to_string();
-
-            document.insert(key, value);
-        }
+        document.insert(key, value);
     }
+
+    // select the attributes to retrieve
+    let attributes_to_retrieve = attributes_to_retrieve
+        .iter()
+        .map(|&fid| field_ids_map.name(fid).expect("Missing field name"));
+
+    let document = permissive_json_pointer::select_values(&document.into(), attributes_to_retrieve);
+
+    // then we need to convert the `serde_json::Map` into an `IndexMap`.
+    let document = document.into_iter().collect();
+
     Ok(document)
 }
 
 fn format_fields<A: AsRef<[u8]>>(
+    document: &Document,
     field_ids_map: &FieldsIdsMap,
-    obkv: obkv::KvReaderU16,
     formatter: &Formatter<A>,
     matching_words: &impl Matcher,
     formatted_options: &BTreeMap<FieldId, FormatOptions>,
 ) -> Result<Document> {
-    let mut document = Document::new();
-
-    for (id, format) in formatted_options {
-        if let Some(value) = obkv.get(*id) {
-            let mut value: Value = serde_json::from_slice(value)?;
-
-            value = formatter.format_value(value, matching_words, *format);
-
-            // This unwrap must be safe since we got the ids from the fields_ids_map just
-            // before.
-            let key = field_ids_map
-                .name(*id)
-                .expect("Missing field name")
-                .to_string();
-
-            document.insert(key, value);
-        }
-    }
+    // Convert the `IndexMap` into a `serde_json::Map`.
+    let document = document
+        .iter()
+        .map(|(k, v)| (k.clone(), v.clone()))
+        .collect();
+
+    let selectors: Vec<_> = formatted_options
+        .keys()
+        // This unwrap must be safe since we got the ids from the fields_ids_map just
+        // before.
+        .map(|&fid| field_ids_map.name(fid).unwrap())
+        .collect();
+
+    let mut document = permissive_json_pointer::select_values(&document, selectors.iter().copied());
+
+    permissive_json_pointer::map_leaf_values(&mut document, selectors, |key, value| {
+        let format = formatted_options
+            .iter()
+            .filter(|(field, _option)| {
+                milli::is_faceted_by(field_ids_map.name(**field).unwrap(), key)
+                    || milli::is_faceted_by(key, field_ids_map.name(**field).unwrap())
+            })
+            .fold(FormatOptions::default(), |acc, (_, option)| {
+                acc.merge(*option)
+            });
```

Comment:
what exactly is going on here? why do we need to merge the format options? can you add some explanations?